File: Python/Graphs/Johnson'sAlgorithm.py
from collections import defaultdict

class Graph:
    def __init__(self, vertices):
        self.V = vertices
        self.graph = defaultdict(list)

    def addEdge(self, u, v, w):
        self.graph[u].append((v, w))

    def BellmanFord(self, src):
        distance = [float("inf")] * (self.V + 1)
        distance[src] = 0

        for _ in range(self.V - 1):
            for u in range(1, self.V + 1):
                for v, w in self.graph[u]:
                    if distance[u] != float("inf") and distance[u] + w < distance[v]:
                        distance[v] = distance[u] + w

        for u in range(1, self.V + 1):
            for v, w in self.graph[u]:
                if distance[u] != float("inf") and distance[u] + w < distance[v]:
                    return None  # Graph contains negative-weight cycle

        return distance[1:]

    def Dijkstra(self, src):
        distance = [float("inf")] * (self.V + 1)
        distance[src] = 0

        visited = [False] * (self.V + 1)

        for _ in range(self.V):
            u = self.minDistance(distance, visited)
            visited[u] = True

            for v, w in self.graph[u]:
                if not visited[v] and distance[u] != float("inf") and distance[u] + w < distance[v]:
                    distance[v] = distance[u] + w

        return distance[1:]

    def minDistance(self, distance, visited):
        minDist = float("inf")
        minDistIndex = -1

        for v in range(1, self.V + 1):
            if not visited[v] and distance[v] <= minDist:
                minDist = distance[v]
                minDistIndex = v

        return minDistIndex

    def Johnson(self):
        # Add a new vertex and connect it to all existing vertices with weight 0
        self.V += 1
        for u in range(1, self.V):
            self.addEdge(self.V, u, 0)

        # Run Bellman-Ford algorithm to get the distances from the extra vertex
        h = self.BellmanFord(self.V)
        if h is None:
            print("Graph contains negative-weight cycles.")
            return

        # Remove the extra vertex and its edges
        self.V -= 1
        self.graph.pop(self.V + 1)

        # Re-weight the edges
        for u in range(1, self.V + 1):
            for i in range(len(self.graph[u])):
                v, w = self.graph[u][i]
                self.graph[u][i] = (v, w + h[u - 1] - h[v - 1])

        # Run Dijkstra's algorithm for each vertex
        for u in range(1, self.V + 1):
            distances = self.Dijkstra(u)
            print(f"Shortest distances from vertex {u}: {distances}")

num=int(input("Enter the number of vertices you want:"))
graph = Graph(num)

for i in range(1,num+1):
  print("Enter source , destination , weight:")
  s,d,w=map(int,input().split())
  graph.addEdge(s,d,w)

# Run Johnson's algorithm
graph.Johnson()

File: C++/Graphs/Johnson'sAlgorithm.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <climits>

using namespace std;

#define INF INT_MAX

struct Edge {
    int source, destination, weight;
};

class Graph {
    int V;
    vector<Edge> edges;

public:
    Graph(int vertices) {
        V = vertices;
    }

    void addEdge(int source, int destination, int weight) {
        Edge edge;
        edge.source = source;
        edge.destination = destination;
        edge.weight = weight;
        edges.push_back(edge);
    }

    vector<int> BellmanFord(int source) {
        vector<int> distance(V, INF);
        distance[source] = 0;

        for (int i = 1; i < V; i++) {
            for (const auto& edge : edges) {
                int u = edge.source;
                int v = edge.destination;
                int weight = edge.weight;
                if (distance[u] != INF && distance[u] + weight < distance[v]) {
                    distance[v] = distance[u] + weight;
                }
            }
        }

        for (const auto& edge : edges) {
            int u = edge.source;
            int v = edge.destination;
            int weight = edge.weight;
            if (distance[u] != INF && distance[u] + weight < distance[v]) {
                // Graph contains negative-weight cycle
                distance.clear();
                return distance;
            }
        }

        return distance;
    }

    vector<int> Dijkstra(int source) {
        vector<int> distance(V, INF);
        distance[source] = 0;

        auto compare = [](pair<int, int>& a, pair<int, int>& b) {
            return a.second > b.second;
        };

        priority_queue<pair<int, int>, vector<pair<int, int>>, decltype(compare)> pq(compare);
        pq.push(make_pair(source, distance[source]));

        while (!pq.empty()) {
            int u = pq.top().first;
            pq.pop();

            for (const auto& edge : edges) {
                if (edge.source == u) {
                    int v = edge.destination;
                    int weight = edge.weight;
                    if (distance[u] != INF && distance[u] + weight < distance[v]) {
                        distance[v] = distance[u] + weight;
                        pq.push(make_pair(v, distance[v]));
                    }
                }
            }
        }

        return distance;
    }

    void Johnson() {
        V++;
        vector<int> h = BellmanFord(V - 1);

        if (h.empty()) {
            cout << "Graph contains negative-weight cycles." << endl;
            return;
        }

        V--;

        for (auto& edge : edges) {
            int u = edge.source;
            int v = edge.destination;
            int weight = edge.weight;
            edge.weight = weight + h[u - 1] - h[v - 1];
        }

        for (int u = 1; u <= V; u++) {
            vector<int> distances = Dijkstra(u);
            cout << "Shortest distances from vertex " << u << ": ";
            for (const auto& dist : distances) {
                if (dist == INF) {
                    cout << "INF ";
                } else {
                    cout << dist << " ";
                }
            }
            cout << endl;
        }
    }
};

int main() {
    int V = 0,s,d,w;
    s=0;
    d=0;
    w=0;
    cout << "Enter the number of vertices: "; 
    cin >> V;
    Graph graph(V);
    for(int i=1;i<=V;i++){
      cout << "Enter the source , destination , weight:";
      cin >> s;
      cin >> d;
      cin >> w;
      graph.addEdge(s,d,w);
    }

    // Run Johnson's algorithm
    graph.Johnson();

    return 0;
}
